### PR TITLE
Enable parallel_map on windows

### LINKED ISSUE #2988: `transpile` is always serial on windows
`qiskit.tools.parallel_map` defaults to running jobs serially whenever the user is running on windows (since the default behavior of `multiprocess.Pool` under windows is considerably different from than that of mac/linux.) However, there are still some cases (like transpiling for RB) where there may be potential benefit to being able to parallelize jobs under windows.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Previously the parallel_map() function, which is used for parallel
execution of tasks such as transpile, was disabled on windows by
default. This was to workaround some issues with launching additional
processes on windows. However, after recent changes like #3491 these
issues should be resolved. This commit enables parallel_map to launch
multiple processes on windows which should speed up execution when
running on windows environments.

### Details and comments

Fixes #2988

### CODE REVIEW

**Comment by mtreinish** [comment]
I expect the jupyter pbars test will still fail here, but once we figure out why that is failing and fix it this should be good to go.

**Comment by nonhermitian** [comment]
I cannot install this PR do to an SSL issue with the `dill` package from `pip`.  However, as a general issue, I always get:
```
BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.
```
on Windows when using `ProcessPoolExecutor`.  Even using `if __name__ == '__main__'` to wrap it.  If instead I use `ThreadPoolExecutor` everything works fine.

This is kind of inline with the idea that Windows is a thread-based OS and does not really understand processes to begin with.

**Comment by mtreinish** [comment]
Hmm, I didn't have any failures when I tested it in my windows vm but I don't think I tested a separate script or jupyter basically just ran the unit tests and everything worked except for the one test. But the jupyter pbars test fails with that same error. I can switch the executor to be the a threadpool for windows if you think that makes sense. I'm not in front of my windows vm right now but I can play with this a bit more when I am

**Comment by nonhermitian** [comment]
I just tried it in a Win 10 VM via Jupyter.  The thing is I am not sure if Threads will give an appreciable speed up.  Starting processes on Windows  requires instantiating a new interpreter each time, which is why that was disabled in the first place.  Did this change so as to make this efficient?

**Comment by mtreinish** [comment]
When I'm back in front of my machine with the windows vm I'll run some numbers. Per #2988 we weren't sure why it was disabled, if it was that the performance was not better and that's still true I'll abandon this and close the issue. But we should generate some numbers and figure out what the performance of threads vs processes vs serial on windows is

**Comment by nonhermitian** [comment]
Yes.  It was the performance.  In most cases firing up a new interpreter is vastly more time consuming than what is typically saved by going to parallel mode.  It is a known issue.  Not sure about threads though.  Typically they execute in parallel, but from what I recall, offer nothing much for cpu intensive performance.

**Comment by mtreinish** [comment]
I finally got a chance to run some numbers on my windows VM (with python 3.7.3). I test the same script with 3 different versions of the parallel_map code: with this PR, without this PR, and with `concurrent.futures.ThreadPoolExecutor` used instead of `ProcessPoolExecutor`. 

The timing data I got was:

```
With this PR:
0:00:15.799924
0:00:17.437665
0:00:17.531656
0:00:15.812527
0:00:15.750009
0:00:16.609192

Without this PR:
0:00:33.484386
0:00:37.218839
0:00:38.421831
0:00:34.796540
0:00:36.343832

With ThreadPoolExecutor:
0:00:53.640598
0:00:53.201676
0:00:54.375017
0:00:58.874668
0:00:52.999554
0:00:55.406509
```

The test script was:
```
from qiskit import *
from qiskit.circuit.random import random_circuit
from qiskit.test import mock
import datetime


if __name__ == '__main__':
    qc = random_circuit(20, 40)
    circuits = [qc] * 4
    start = datetime.datetime.utcnow()
    transpile(circuits, backend=mock.FakeTokyo())
    stop = datetime.datetime.utcnow()
    print(stop - start)
```

Based on these numbers it looks like turning on parallel execution with processes does offer a noticeable speedup.

As for the error you were hitting I was able to avoid it in a standalone script with `if __name__ == '__main__'`. (I'll add a release note here about this) I still haven't tested things inside jupyter yet.

**Comment by nonhermitian** [comment]
Ok cool. The name main trick did not work for me in jupyter. But also if one has to do this everywhere that parallel is called it will break alot of stuff

**Comment by mtreinish** [comment]
I'm marking this as on hold because I think we should wait for #5324 to merge first. That actually implicitly fixes this because it removes windows as being a hard coded skip for `parallel_map`. But makes it opt-in for windows instead of on by default. After that merges I'll rebase this and add a release note about running in parallel on windows and the limitation with it.

File: releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml
Comment (kdk) [inline releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml:15]:
```suggestion
    parallelism may not work when called from a script running outside of a
```

File: releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml
Comment (kdk) [inline releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml:9]: Do you think it'd be beneficial to direct the user to the issue note here? (They won't necessarily appear in this order in the final release notes, right?)

File: releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml
Comment (kdk) [inline releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml:20]: I found this sentence a little hard to parse. The issue can occur when a runtime or attribute error is raised from code called through parallel_map, or code calling parallel_map?

**Review by kdk — COMMENTED** [review]
This looks good. Can you open an issue on the metarepo to include in the user config docs a note on the required block for running in parallel on windows?

File: releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml
Comment (mtreinish) [reply releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml:9]: Yeah, that's a good idea I'll leave a note about there being potential issues and refer to the known issues section

File: releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml
Comment (mtreinish) [reply releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml]: Yeah it was worded poorly, I was trying to say anywhere parallel_map is used which included things calling it internally, like transpile(). I've updated it in: https://github.com/Qiskit/qiskit-terra/pull/3547/commits/44e5ebceed55a70a52ae9bfc321e72e9a4ba486e to try and clean it up.

File: releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml
Comment (mtreinish) [reply releasenotes/notes/enable-windows-parallel-map-b33122a2122a95d7.yaml]: Done in: https://github.com/Qiskit/qiskit-terra/pull/3547/commits/44e5ebceed55a70a52ae9bfc321e72e9a4ba486e

**Review by kdk — COMMENTED** [review]
Tests erroring in Windows CI with `AttributeError: module 'multiprocessing.context' has no attribute 'ForkProcess'`.

**Comment by mtreinish** [comment]
hmm I guess I actually needed: https://github.com/Qiskit/qiskit-terra/pull/3547/commits/7ee74becd0de83454bd8f7e844cb3f80432cb6f5 looking at the code (which I wrote > 1 year ago) it wasn't obvious to me why that was needed. But I guess it is because on windows they don't even bother to define fork process.

**Review by kdk — APPROVED** [review]

**Pull request merged**
